tests/test_server.py

```diff
@@ -30,20 +30,32 @@
 
 try:
     # Failing to import any of these should disable the tests.
-    import lsst.daf.butler.server
     from fastapi.testclient import TestClient
-    from lsst.daf.butler.server import app
+    from lsst.daf.butler.remote_butler import RemoteButler
+    from lsst.daf.butler.remote_butler.server import Factory, app, factory_dependency
 except ImportError:
     TestClient = None
     app = None
 
-from lsst.daf.butler import Butler, CollectionType, Config, DataCoordinate, DatasetRef
-from lsst.daf.butler.tests import addDatasetType
+from lsst.daf.butler import Butler
 from lsst.daf.butler.tests.utils import MetricTestRepo, makeTestTempDir, removeTestTempDir
 
 TESTDIR = os.path.abspath(os.path.dirname(__file__))
 
 
+def _make_remote_butler(http_client):
```

Comment:
This is not how a normal user is going to create a RemoteButler. They are going to say:

```python
butler = Butler("https://test.example")
```
and it is meant to work automatically. It's meant to automatically look for the `/butler.json` endpoint,

It's this code in the prototype server:

```python
app.get("/butler/butler.json", response_model=dict[str, Any])
 def read_server_config() -> Mapping:
     """Return the butler configuration that the client should use."""
     config_str = f"""
 datastore:
     root: {BUTLER_ROOT}
 registry:
     cls: lsst.daf.butler.registries.remote.RemoteRegistry
     db: <butlerRoot>
 """
     config = Config.fromString(config_str, format="yaml")
     return config.toDict()
```

This would clearly need to change to return back the `cls`. There is a `<butlerRoot>` convention in the configs that allows the config to say "use what ever URI you found this butler config at" versus a hard-coded server name. So if you go to `https://192.168.1.1/butler.json` and the server is at `https://192.168.1.1` that all works if you set the `root` field to `<butlerRoot>` and if you don't like `root` we can call it `server_url` or something. Alternatively the butler.json could return a hard-coded server URL that is distinct from where the config came from.

In the test code below that was deleted you will see that it downloads the Config, then injects the httpx client into the config such that the Butler constructor will get the Client object for the server URI rather than having to make it itself from a URI.

We can leave it like this for now but it needs to be fixed to work the standard way soon. Remember that an end user doesn't have to know whether they ended up with a DirectButler or RemoteButler.

Reply:
By convention in Phalanx/Safir for REST APIs they don't allow external access to things at the top level of the HTTP path -- that's reserved for internal endpoints like Kubernetes health checks.  The externally accessible parts are expected to be nested under "/butler/" or something.  Not sure if we can override that or not, I can find out.

As it stands you can point the constructor to a config file on disk or an HTTP server and get back a remote butler instance the same way as you would with a direct butler.  Whether it makes sense to serve that file from the REST server itself I'm not sure.  If it does it may be more like "http://butler.lsst.cloud/configs/dp02" rather than just the root URL.  

I kind of think that any configuration chunks that must be provided by the server should come from a versioned endpoint instead of a static file.  So if we are keeping the convention from DirectButler that you can load a configuration file from an http server, we might want to do like "butler://butler.lsst.cloud" to mean "create a remote butler pointing at butler.lsst.cloud with all client-side options set to defaults"